### PR TITLE
Add clarity to slot arithmetic functions.

### PR DESCRIPTION
**Pull request by jonathanknowles**
# Issue Number

Indirectly related to #466.

# Overview

This PR:

- [x] refactors existing slot arithmetic functions to make them easier to call, and to make their behaviour more obvious.
- [x] adds a set of tests to verify that these refactorings are correct.
- [x] adds a small set of tests to test the previously-untested `slotDifference` function.

### CODE REVIEW

File: lib/core/src/Cardano/Wallet/Primitive/Types.hs [hunk 1]
```diff
@@ -876,32 +892,53 @@ slotDifference epl a b
     a' = flatSlot epl a
     b' = flatSlot epl b
 
+-- | Return the slot immediately before the given slot.
```
Comment (KtorZ) [inline lib/core/src/Cardano/Wallet/Primitive/Types.hs:895]: Gotta admit that I am not fond of the alignment style below :/ ... Plus it clashes with our policy of no variable-length indentation!

File: lib/core/src/Cardano/Wallet/Primitive/Types.hs [hunk 2]
```diff
@@ -876,32 +892,53 @@ slotDifference epl a b
...
+slotPred :: SlotParameters -> SlotId -> SlotId
+slotPred (SlotParameters (EpochLength el) _ _) (SlotId en sn)
+    | sn > 0    = SlotId (en    ) (               sn - 1)
+    | otherwise = SlotId (en - 1) (fromIntegral $ el - 1)
+
+-- | Return the slot immediately after the given slot.
+slotSucc :: SlotParameters -> SlotId -> SlotId
+slotSucc (SlotParameters (EpochLength el) _ _) (SlotId en sn)
+    | sn < el - 1 = SlotId (en    ) (sn + 1)
+    | otherwise   = SlotId (en + 1) (     0)
```
Comment (KtorZ) [inline lib/core/src/Cardano/Wallet/Primitive/Types.hs]: What's the need for these two functions :thinking: ? They are unused and weren't needed before. 
Truth is, I'd rather not have to maintain code that we don't use and make the codebase bigger than necessary.

**Review by KtorZ — CHANGES_REQUESTED** [review]
Hmmm... Good refactor, though it also introduces needless functions that can be avoided: 

we don't know what the future is made of, so let's not add things that might possibly be used. Let's add them when needed.

**Review by paweljakubas — APPROVED** [review]
LGTM except probably not needed at this moment (slotSucc)

File: lib/core/src/Cardano/Wallet/Primitive/Types.hs [hunk 3]
```diff
@@ -876,32 +892,53 @@ slotDifference epl a b
...
-        { epochNumber = ep
-        , slotNumber = sl
-        }
+    offset = sl * fromIntegral (flatSlot el slot)
+
+-- | For the given time 't', determine the ID of the earliest slot with start
+--   time 's' such that 't ≤ s'.
+slotStartingAtOrJustAfter :: SlotParameters -> UTCTime -> SlotId
+slotStartingAtOrJustAfter sp@(SlotParameters _ (SlotLength sl) _) t =
+    slotAt sp (addUTCTime (pred sl) t)
```
Comment (paweljakubas) [inline lib/core/src/Cardano/Wallet/Primitive/Types.hs:918]: we want to use `slotPred`, right?

File: lib/core/src/Cardano/Wallet/Primitive/Types.hs [hunk 4]
```diff
@@ -876,32 +892,53 @@ slotDifference epl a b
     a' = flatSlot epl a
     b' = flatSlot epl b
 
+-- | Return the slot immediately before the given slot.
+slotPred :: SlotParameters -> SlotId -> SlotId
```
Comment (paweljakubas) [inline lib/core/src/Cardano/Wallet/Primitive/Types.hs:896]: function is not used except tests I believe

File: lib/core/src/Cardano/Wallet/Primitive/Types.hs (hunk 1 above)
Comment (jonathanknowles) [reply lib/core/src/Cardano/Wallet/Primitive/Types.hs:895]: > Gotta admit that I am not fond of the alignment style below :/ ... Plus it clashes with our policy of no variable-length indentation!

You're right! I'll remove this indentation.

(I was using this style while writing as it made it visually easier for me to read, but forgot to remove it before pushing.)

File: lib/core/src/Cardano/Wallet/Primitive/Types.hs (hunk 2 above)
Comment (jonathanknowles) [reply lib/core/src/Cardano/Wallet/Primitive/Types.hs]: > What's the need for these two functions thinking ? They are unused and weren't needed before.

Their existence makes it easier to express the following two laws:
```hs
slotSucc . slotStartingAtOrJustBefore . utcTimePred . slotStartTime == id
slotPred . slotStartingAtOrJustAfter  . utcTimeSucc . slotStartTime == id
```

Without `slotPred` and `slotSucc`, we could use `slotDifference`. However, _that_ function doesn't appear to be tested: we might also want to write a test for this function.

File: lib/core/src/Cardano/Wallet/Primitive/Types.hs (hunk 3 above)
Comment (jonathanknowles) [reply lib/core/src/Cardano/Wallet/Primitive/Types.hs:918]: > we want to use `slotPred`, right?

I'm not sure that would work. What would you suggest?

These functions are connected with the following laws:
```hs
slotSucc . slotStartingAtOrJustBefore . utcTimePred . slotStartTime == id
slotPred . slotStartingAtOrJustAfter  . utcTimeSucc . slotStartTime == id

slotStartingAtOrJustBefore . slotStartTime == slotStartingAtOrJustAfter . slotStartTime
```

File: lib/core/src/Cardano/Wallet/Primitive/Types.hs (hunk 4 above)
Comment (jonathanknowles) [reply lib/core/src/Cardano/Wallet/Primitive/Types.hs:896]: > function is not used except tests I believe

That's correct. However, these functions make it easier to express the following laws:
```hs
slotSucc . slotStartingAtOrJustBefore . utcTimePred . slotStartTime == id
slotPred . slotStartingAtOrJustAfter  . utcTimeSucc . slotStartTime == id
```

File: lib/core/src/Cardano/Wallet/Primitive/Types.hs (hunk 1 above)
Comment (jonathanknowles) [reply lib/core/src/Cardano/Wallet/Primitive/Types.hs:895]: Fixed.

**Comment by jonathanknowles** [comment]
> we don't know what the future is made of, so let's not add things that might possibly be used.
> Let's add them when needed.

These functions are already used!

Their existence makes it convenient to express the following two laws, which are tested:
```hs
slotSucc . slotStartingAtOrJustBefore . utcTimePred . slotStartTime == id
slotPred . slotStartingAtOrJustAfter  . utcTimeSucc . slotStartTime == id
```

**Review by Anviking — APPROVED** [review]
Nice, lgtm! Maybe the functions used just for tests should be moved to the tests.

File: lib/core/src/Cardano/Wallet/Primitive/Types.hs (hunk 2 above)
Comment (jonathanknowles) [reply lib/core/src/Cardano/Wallet/Primitive/Types.hs]: @KtorZ To that end, I've added a small set of tests for `slotDifference` (as far as I can tell, it was previously untested, at least not directly).

**Comment by paweljakubas** [comment]
maybe it is possible to use those functions somewhere in code as a invariant???

**Comment by jonathanknowles** [comment]
@paweljakubas wrote:

> maybe it is possible to use those functions somewhere in code as a invariant???

Interesting. Could you give an example?

**Review by KtorZ — APPROVED** [review]

**Pull request merged**